kb/2024-01-30/the_potential_risk_with_fstrim.md

```diff
@@ -0,0 +1,65 @@
+---
+title: The potential risk with fstrim
+description: The potential risk with fstrim and how to avoid it
+slug: the_potential_risk_with_fstrim
+authors:
+  - name: Vicente Cheng
+    title: Senior Software Engineer
+    url: https://github.com/Vicente-Cheng
+    image_url: https://github.com/Vicente-Cheng.png
+tags: [harvester, rancher integration, longhorn, fstrim]
+hide_table_of_contents: false
+---
+
+The `fstrim` is the common way to release the unused space of the filesystem. However, we encounter the known issue with `fstrim` on the Longhorn volume. This article shares the potential risk with `fstrim` and how to avoid it.
+
+The known issue is that executing the `fstrim` on the Longhorn volume may result in IOErrors if the volume is rebuilding. Related issue: (You can find more details in the issues)
+  - https://github.com/harvester/harvester/issues/4739
+  - https://github.com/longhorn/longhorn/issues/7103
+
+## The potential risk and affection with fstrim
+
+If you encounter the known issue on the above, that will result in the IOErrors. The IOErrors will cause the pod that uses this volume to be stuck. If the pod is critical, it will cause the application to be unavailable. For example, Harvester usually uses the Longhorn volume as the VM disk. After encountering this issue, the VM will flap in pause and running state until the volume rebuild is completed.
+
+That does not affect the data integrity, but it will cause some panic issues for users. It caused the VM to hang, and the application will be unavailable. Consider the guest Kubernetes cluster scenario. When the VM is unavailable, it means the etcd service is not available. If half of the etcd service is unavailable, the Kubernetes cluster will be unavailable. Meanwhile, any services running on this Kubernetes cluster will be unavailable.
+
+## How to avoid the potential risk
+
+The way to avoid the potential risk is to disable the `fstrim`. The `fstrim` is enabled by default on various modern Linux distributions.
+You can check the following items for the potential `fstrim`.
+  - Disable the `fstrim` by adding the `nodiscard` option to the mount options on the `/etc/fstab` file. (`nodiscard` sometimes is the default value, but you can still add it.)
+  
+    You can refer to the following cloud-init template.
```

Comment:
Can we first have an example on how to change the fstab file, and then followed by the cloud-init example to configure the fstab while creating a VM.

Reply:
Remove the fstab part. Because the mount does not have an option called `nodiscard,` the default is to discard; we need to use the option `discard` if we want.